secondshape.go

```diff
@@ -0,0 +1,93 @@
+package retro
+
+import (
+	"errors"
+	"time"
+)
+
+// ErrorBanAttemptsReached occurs when the execute function was executed BanAttempts times
+var ErrorBanAttemptsReached = errors.New("BanAttempts reached")
+
+// ErrorShouldRetryFunctionError occurs when ShouldRetry function returns false
+var ErrorShouldRetryFunctionError = errors.New("ShouldRetry returned false")
+
+// ErrorRecoveryFunctionError occurs when Recovery function returns an error
+var ErrorRecoveryFunctionError = errors.New("Recovery function finished with error")
+
+// ErrorExecuteFunctionNil occurs when Run() is executed without Execute function defined
+var ErrorExecuteFunctionNil = errors.New("Execute function has not beed defined")
+
+// ErrorMaxAttemptsReached occurs when the execute function was executed MaxAttempts times
+var ErrorMaxAttemptsReached = errors.New("MaxAttempts reached")
+
+// ErrorMaxAttemptsIsZero occurs as Max Attempts is supposed to be > 0
+var ErrorMaxAttemptsIsZero = errors.New("Max Attempts is zero")
+
+// ErrorDelayIsZero occurs as Delay is supposed to be > 0
+var ErrorDelayIsZero = errors.New("Delay is zero")
+
+// ErrorBanTimeoutIsZero occurs as Ban Timeout is supposed to be > 0
+var ErrorBanTimeoutIsZero = errors.New("Ban Timeout is zero")
+
+// Execute function
+type Execute func() error
+
+// Recovery function
+type Recovery func(error) error
+
+// ShouldRetry function
+type ShouldRetry func(error) bool
+
+type strategy interface {
+	getExecute() (Execute, error)
```

Comment:
This interface could probably be simplified a bit.

Also Im not sure strategy should have anything to do with execution. Strategy should probably define the behavior in response to errors, so all strategy should need access to is the error being retried and the configuration of the strategy. You could also provide state around the number of attempts to help provide context to help inform the decision.

This could be simplified a lot by driving these decisions inline with the error handling, so where strategy has no concept of the execution or the state itself. In this case you'd have `Try(myfunc, BasicStrategy)` where the logic around retrying would maintain state, and poll parameters of the strategy to decide how to handle errors. 

Something like

```
type state struct {
  numAttempts int
}

type strategy interface {
  shouldRetry func(error, state) bool
  retryBackoff(state) time.Duration
}
```